### PR TITLE
Hyper7

### PR DESCRIPTION
**Pull request by bit-shifter**
## Description

Adds both a default and user keymap for the hyper7 v3 PCB.

Contains 2 layouts: default and southpaw.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by keyboard-magpie** [comment]
Have tested all aspects of the new keymaps on macOs, happy to say they work.

I  don't know if it's worth including in the readme about unicode input modes on win/macos?

**Comment by bit-shifter** [comment]
I've noticed a rather large problem with the latest commit, working to resolve.

Removing the additional keys from the layout macro to satisfy the `info.json`'s 183 key representation has caused the board to misbehave.

Always verify changes before pushing!

File: keyboards/hyper7/h7/v3/config.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+#define CONFIG_H
+
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/config.h:3]:
```suggestion

```

File: keyboards/hyper7/h7/v3/config.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xDEAD
+#define PRODUCT_ID      0x6060
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    yiancar
+#define PRODUCT         hyper7 v3.0
+#define DESCRIPTION     hyper7 v3.0 firmware compiled by bit-shifter
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/config.h:12]:
```suggestion
```

File: keyboards/hyper7/h7/v3/config.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
+
+// see: https://docs.qmk.fm/#/feature_leader_key
+#define LEADER_PER_KEY_TIMING
+#define LEADER_TIMEOUT 250
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/config.h:44]:
```suggestion
```

File: keyboards/hyper7/h7/v3/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,760 @@
+#include QMK_KEYBOARD_H
+
+#include "../../config.h"
+#include "../../../../../../quantum/quantum.h"
+#include "../../../../../../quantum/quantum_keycodes.h"
+
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/keymaps/default/keymap.c:6]:
```suggestion

```

File: keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c [hunk 5]
```diff
@@ -0,0 +1,725 @@
+#include QMK_KEYBOARD_H
+
+#include "../../config.h"
+#include "../../../../../../quantum/quantum_keycodes.h"
+#include "bit-shifter.h"
+
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c:6]:
```suggestion

```

File: keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+        FT,      FT,      FT,       FT,       FT,       FT,       FT,       FT,       UC_M_WI,  FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,      FT,      FT,       FT,                            FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       FT,       FT,       UC_MOD,   FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       UC_M_LN,  FT,       FT,       FT,      FT,      FT,                                      FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       H_LOCK,   FT,       FT,       FT,       FT,       FT,       FT,       UC_M_WC,  FT,       FT,       FT,       UC_M_MA,  FT,       FT,       FT,       FT,      FT,      FT,       FT,                            FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,      FT,      FT,                                      FT,      FT,       FT,      FT
+    )
+};
+
+void matrix_init_keymap(void) {}
+
+void matrix_scan_keymap(void) {}
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c:495]:
```suggestion

```

File: keyboards/hyper7/h7/v3/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+        FT,      FT,      FT,       FT,       FT,       FT,       FT,       FT,       UC_M_WI,  FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,      FT,      FT,       FT,                            FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       FT,       FT,       UC_MOD,   FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       UC_M_LN,  FT,       FT,       FT,      FT,      FT,                                      FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       H_LOCK,   FT,       FT,       FT,       FT,       FT,       FT,       UC_M_WC,  FT,       FT,       FT,       UC_M_MA,  FT,       FT,       FT,       FT,      FT,      FT,       FT,                            FT,      FT,       FT,      FT,
+        FT,      FT,      FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,       FT,      FT,      FT,                                      FT,      FT,       FT,      FT
+    )
+};
+
+void matrix_init_keymap(void) {}
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/keymaps/default/keymap.c:501]:
```suggestion
```

File: keyboards/hyper7/h7/v3/rules.mk [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+OPT_DEFS += -DBOOTLOADER_SIZE=8192
+
+
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/rules.mk:43]:
```suggestion

# Bootloader selection
BOOTLOADER = atmel-dfu

```

File: keyboards/hyper7/h7/v3/rules.mk [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+CONSOLE_ENABLE = no	     # Console for debug
+COMMAND_ENABLE = no      # Commands for debug and configuration
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+BACKLIGHT_ENABLE = no    # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no     # Enable keyboard RGB underglow
+AUDIO_ENABLE = no        # Audio output
+
+LAYOUTS = h7v3_183 h7v3_183_southpaw h7v3_213 h7v3_213_southpaw
```
Comment (zvecr) [inline keyboards/hyper7/h7/v3/rules.mk:64]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Also, all `*.h` and `*.c` files need a GPL compatible header added.

File: users/bit-shifter/bit-shifter.c [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+#ifdef KEYLOGGER_ENABLE
+  xprintf("KL: row: %u, column: %u, pressed: %u\n", record->event.key.col, record->event.key.row, record->event.pressed);
+#endif //KEYLOGGER_ENABLE
+
+  return process_record_keymap(keycode, record);
+}
+
+// Any custom LED code goes here.
+// Currently just proxies to `led_set_keymap`
+void led_set_user(uint8_t usb_led) {
```
Comment (fauxpark) [inline users/bit-shifter/bit-shifter.c:78]: Please use `led_update_*()` if possible: https://docs.qmk.fm/#/feature_led_indicators

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by keyboard-magpie** [comment]
Oopsy. I'll go and nudge @bit-shifter in real life.

File: keyboards/hyper7/h7/v3/config.h (hunk 1 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/config.h:3]: Thanks - removed in local fork

File: keyboards/hyper7/h7/v3/config.h (hunk 2 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/config.h:12]: Removed - is there any need for the description field?

File: keyboards/hyper7/h7/v3/config.h (hunk 3 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/config.h:44]: Why should these be removed?

There's leader key functionality defined in my users' settings, which defines some example multi-key examples.

File: keyboards/hyper7/h7/v3/keymaps/default/keymap.c (hunk 4 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/keymaps/default/keymap.c:6]: These were added to help with local development - happy to remove, but could someone point me in the direction of some documentation showing how to set up a local env?

File: keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c (hunk 5 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c:6]: Same as previous comment, will remove.

File: keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c (hunk 6 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/keymaps/bit-shifter/keymap.c:495]: Removed, thanks

File: keyboards/hyper7/h7/v3/keymaps/default/keymap.c (hunk 7 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/keymaps/default/keymap.c:501]: Removed

File: keyboards/hyper7/h7/v3/rules.mk (hunk 9 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/rules.mk:64]: I'm unsure on the request to replace the current rule flags (which seem necessary) with these.

Specifically: LEADER_ENABLE, KEY_LOCK_ENABLE and the list of LAYOUTS

File: users/bit-shifter/bit-shifter.c (hunk 10 above)
Comment (bit-shifter) [reply users/bit-shifter/bit-shifter.c:78]: Updated, although I'm unsure whether the approach taken is valid. It's currently a stub impl.

File: keyboards/hyper7/h7/v3/rules.mk (hunk 8 above)
Comment (bit-shifter) [reply keyboards/hyper7/h7/v3/rules.mk:43]: I've made the suggested change here, but would like confirmation that they're equivalent.

**Comment by bit-shifter** [comment]
Changes pending, keen to reopen this pull request to keep track of comment history

**Comment by bit-shifter** [comment]
Opened a new PR: #12793.

More than happy to close and revert to this one if a maintainer is happy to reopen

**Pull request closed without merging**